axelrod/strategies/titfortat.py

```diff
@@ -92,6 +93,43 @@ class TwoTitsForTat(Player):
     def strategy(opponent: Player) -> Action:
         return D if D in opponent.history[-2:] else C
 
+class DynamicTwoTitsForTat(Player):
+    """
+    A player starts by cooperating and then punishes its opponent's 
+    defectiions by opponent with a dynamic bias based off of the 
+    opponents ratio of cooperations to total moves (so their current 
+    probability of cooperating towards cooporating regardless of the 
+    move (aka: forgiveness)).
+    Names:
+
+     - Dynamic Two Tits For Tat: Original name by Grant Garrett-Grossman.
+    """
+
+    name = 'Dynamic Two Tits For Tat'
+    classifier = {
+        'memory_depth': 2,  # Long memory, memory-2
+        'stochastic': True,
+        'makes_use_of': set(),
+        'long_run_time': False,
+        'inspects_source': False,
+        'manipulates_source': False,
+        'manipulates_state': False
+    }
+    
+    @staticmethod
+    def strategy(opponent):
+		# First move
+        if len(opponent.history) == 0:
+            # Make sure we cooporate first turn
+            return C
+        if D in opponent.history[-2:]:
+            # Probability of cooperating regardless
+            if random_choice(opponent.cooperations / len(opponent.history)):
```

Comment:
`random_choice` does not return a boolean but returns either a `C` or a `D`. This should be:

```python
125         if D in opponent.history[-2:]:                                                                                                          
126             # Probability of cooperating regardless                                                                                             
127             return random_choice(opponent.cooperations / len(opponent.history))                                                                 
128         return C     
```

I have checked this locally and it all works, you'll need to adjust the tests and the actions. For example the following are passing tests for this strategy:

```python
134         opponent = axelrod.MockPlayer(actions=[D, C, D, D, C])                                                                                  
135         actions = [(C, D), (D, C), (C, D), (D, D), (D, C)]                                                                                      
136         self.versus_test(opponent, expected_actions=actions, seed=1)                                                                            
137                                                                                                                                                 
138         actions = [(C, D), (D, C), (D, D), (D, D), (C, C)]                                                                                      
139         self.versus_test(opponent, expected_actions=actions, seed=2) 
```

As well as those two I'd just add tests with the `Defector` and `Cooperator` (the case when this strategy does not act randomly):

```python
140                                                                                                                                                 
141         actions = [(C, C), (C, C), (C, C), (C, C), (C, C)]                                                                                      
142         self.versus_test(axelrod.Cooperator(), expected_actions=actions)                                                                        
143                                                                                                                                                 
144         actions = [(C, D), (D, D), (D, D), (D, D), (D, D)]                                                                                      
145         self.versus_test(axelrod.Defector(), expected_actions=actions)  
```